nengo_spa/algebras/base.py

```diff
@@ -220,6 +301,66 @@ def implement_binding(self, n_neurons_per_d, d, unbind_left, unbind_right):
         """
         raise NotImplementedError()
 
+    def sign(self, v):
+        """Returns the sign of *v* defined by the algebra.
+
+        The exact definition of the sign depends on the concrete algebra, but
+        should be analogous to the sign of a number in so far that
+
+        * binding of two "positive" vectors produces another "positive" vector,
+        * binding of a "negative" vector with itself produces a "positive"
+          vector,
+        * binding of a "negative" and "positive" vectors produces a "negative"
+          vector.
+
+        Furthermore, if the algebra supports fractional binding powers, it
+        should do so for all "non-negative" vectors, but not "negative" vectors.
+
+        If an algebra does not have the notion of a sign, it may raise a
+        *NotImplementedError*.
+
+        Parameters
+        ----------
+        v : (d,) ndarray
+            Vector to determine sign of.
+
+        Returns
+        -------
+        AbstractSign
+            The sign of the input vector.
+
+        See Also
+        --------
+        AbstractAlgebra.abs
+        """
+        raise NotImplementedError()
+
+    def abs(self, v):
+        """Returns the absolute vector of *v* defined by the algebra.
```

Comment:
One more thought: since the property `UNITARY` has the method `make_unitary`, maybe the property `POSITIVE` could have the method `make_positive` (instead of `abs`)?

Reply:
Essentially for the reasons described in the PR description under "Vector creation API". It does not really scale well to add a new `make_something` method for any potential property (and potentially the cartesian product of these properties). Also, `abs` is more analogous to other APIs like `np.abs`.

Reply:
Yeah I read that part and this isn't really about scaling with the cartesian product. It's more that I'm not sure of that analogy. For instance `np.abs(-3 - 4j) = 5` but the HRR abs as it is defined here only changes the sign of 1 or 2 coefficients while leaving the rest unchanged. So I'm not sure if this analogy helps avoid confusion whereas `POSITIVE` is nice because it is more clearly referring to a property that is defined by the algebra.

Reply:
I was thinking that if you could implement it like:
```
def abs(self, v):
    if self.sign(v).is_negative():
        v = self.negate(v)
    return v
```
where
```
def negate(self, v):
    return self.bind(self.negative_identity_element(len(v)), v)
```
then I would see this as indeed being very much analogous to `abs`. I tried this out and two of the HRR unit tests failed. Not sure why or if that's fixable without any negative consequences? I don't think so because the vector that needs to be binded with `v` is context-sensitive, in that there are 3 different choices corresponding to the ones in `HRRSign.to_vector`.

Reply:
Actually scratch that, I don't think my last post is a good argument at all because you could fix `negate` by defining the LHS term in a way that is context sensitive in `v`, and there's no good reason to require a priori that `abs` should be defined in this way. The example I gave of `np.abs(-3 - 4j) == 5` is an example where `abs` cannot be defined in such a way.

Perhaps the part that is making me uncomfortable is there's generally only one way to flip the sign of something, and I believe that is usually consistent with multiplying by the negative identity. But here there are two ways to make something negative. And to make something positive, the thing that you have to multiply it with varies depending on which way it is negative. So I guess I'm having a hard time making these analogies fully consistent with one another in my head right now.

Reply:
Okay I think I've figured out what was making me uncomfortable and how to fix it. I see the analogy to `abs` now as being derived from this definition of sign:
```
sign(z) = z / abs(z)
```
which is exact same definition as the one for complex numbers, and could also be the definition here. Indeed, if you multiply both sides by `abs(z)` then you have:
```
abs(z) = (1 / sign(z)) * z
```
where `1 / sign(z)` is consistent with `HRRSign.to_vector` because `1 / -1 == -1 ` and `1 / 1 == 1`. And `*` is binding. This is currently the base implementation for `abs`! Hurray.

Somewhat different from how I had been thinking about this until now, but I think that actually makes the analogy more complete and consistent in combination with [the comment about multiplying signs](https://github.com/nengo/nengo-spa/pull/271/files#r595706491)...?

Reply:
I wasn't even aware of the `sign` definition for complex numbers, but it seems to fit nicely.